res/controllers/midi-components-0.1.js

```diff
@@ -0,0 +1,401 @@
+/**
+ * Components JS library for Mixxx
+ * Documentation is on the Mixxx wiki at
+ * https://github.com/mixxxdj/mixxx/wiki/Components-JS
+ *
+ * Copyright (C) 2017 Be <be.0@gmx.com>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+(function(global) {
+    /**
+     * @typedef ComponentOpts
+     * @property {number=} midiStatus
+     * @property {number=} midiNo
+     * @property {string=} group
+     * @property {string=} inKey
+     * @property {string=} outKey
+     * @property {boolean} [outConnect=true]
+     * @property {boolean} [outTrigger=true]
+     * @property {number} [max=127]
+     * @property {number} [shiftOffset=0]
+     * @property {Component.SendShifted} [sendShifted=Component.SendShifted.NO]
+     */
+    class Component {
+        /** @param {ComponentOpts} opts */
+        constructor({
+            midiStatus = undefined,
+            midiNo = undefined,
+            group = undefined,
+            inKey = undefined,
+            outKey = undefined,
+            outConnect = true,
+            outTrigger = true,
+            max = 127,
+            shiftOffset = 0,
+            sendShifted = Component.SendShifted.NO
+        } = {}) {
+            /**
+             * @type {ScriptConnection[]}
+             * @protected
+             */
+            this._connections = [];
+            this.isShifted = false;
+
+            this.midiStatus = midiStatus;
+            this.midiNo = midiNo;
+            this.group = group;
+            this.inKey = inKey;
+            this.outKey = outKey;
+            this.outConnect = outConnect;
+            this.outTrigger = outTrigger;
+            this.max = max;
+            this.shiftOffset = shiftOffset;
+            this.sendShifted = sendShifted;
+
+            if (this.outConnect) {
+                this.connect();
+                this.outTrigger ? this.trigger() : this.noop();
+            }
```

Comment:
This strictly typed approach looses the feature where the user can add extra properties to the type when instantiating it. I think that is one of the strengths in QML and components-0.0. I fear the usage in mappings will get too clunky if we remove that feature. See [my reply on zulip](https://mixxx.zulipchat.com/#narrow/stream/113295-controller-mapping/topic/Pioneer.20DDJ-SR2/near/448621304) where I listed a bunch of usage examples where I used that pattern to avoid more broadly scoped (global) variables.

Reply:
IMO, using the constructor parameter to override any component method defeats the purpose of using a class system to easier IDEs code completion and type hinting. If a user wants to override a component's methods without explicitly declaring a class, they can write the following:

`js
const comp = new class extends components.Component {
    inValueScale(value) {
        return value === 0x40 ? true : false;
    }
}({midiStatus: 0x70, midiNo: 0x40}
```